### PR TITLE
New hash format

### PR DESCRIPTION
**Pull request by e-ago**
New hash format supported

### CODE REVIEW

**Comment by kholia** [comment]
The code is looking cleaner now. Good job!

1. The algorithm is not `PBKDF2-SHA256 AES OpenCL`. Please change it to `Iterated-SHA256 AES OpenCL` or `SHA256 AES OpenCL`. I like the `Iterated-SHA256` version since it is a bit more informative.

2. You can reuse the `bitlocker_tests` which are already defined in the `bitlocker_common_plug.c` file. This is not a big problem and the decision is left to you.

**Comment by solardiz** [comment]
On Fri, Apr 21, 2017 at 11:24:18PM -0700, Dhiru Kholia wrote:
> 1. The algorithm is not `PBKDF2-SHA256 AES OpenCL`. Please change it to `Iterated-SHA256 AES OpenCL` or `SHA256 AES OpenCL`. I like the `Iterated-SHA256` version since it is a bit more informative.

This is indeed just a bikeshed, but FWIW we haven't been including this "Iterated-" thing in other formats - e.g., descrypt is also iterated (25 iterations), but we do not say this anywhere in the algorithm name string (we merely list "DES" as a primitive used).  The way we've been using the algorithm string so far is to list the names of the primitives used, so it would be just "SHA-256 AES" here (well, and "OpenCL", although it's not quite in the same category - but we've been doing that).  Also, "SHA-256" is normally spelled with the dash in it (the dash gets dropped in "HMAC-SHA256", because that's how others spell those HMACs, although I think we're now inconsistent about that).

**Comment by kholia** [comment]
OK.

I will apply the following patch to the BitLocker CPU format,

```
$ git diff
diff --git a/src/bitlocker_fmt_plug.c b/src/bitlocker_fmt_plug.c
index 7aae5d0..5d577d6 100644
--- a/src/bitlocker_fmt_plug.c
+++ b/src/bitlocker_fmt_plug.c
@@ -46,9 +46,9 @@ john_register_one(&fmt_bitlocker);
 
 #define FORMAT_LABEL            "BitLocker"
 #if ARCH_BITS >= 64
-#define ALGORITHM_NAME          "Iterated-SHA256 AES 64/" ARCH_BITS_STR
+#define ALGORITHM_NAME          "SHA-256 AES 64/" ARCH_BITS_STR
 #else
-#define ALGORITHM_NAME          "Iterated-SHA256 AES 32/" ARCH_BITS_STR
+#define ALGORITHM_NAME          "SHA-256 AES 32/" ARCH_BITS_STR
 #endif
 #define BENCHMARK_COMMENT       ""
 #define BENCHMARK_LENGTH        -1
```
Thanks!

File: src/opencl_bitlocker_fmt_plug.c [hunk 1]
```diff
@@ -358,9 +353,9 @@ static void done(void)
 
 static void *get_salt(char *ciphertext)
 {
-	static char dummy[BITLOCKER_JTR_HASH_SIZE_CHAR + 1];
+	static char dummy[PLAINTEXT_LENGTH + 1];
 
-	return strnzcpy(dummy, ciphertext, BITLOCKER_JTR_HASH_SIZE_CHAR + 1);
+	return strnzcpy(dummy, ciphertext, PLAINTEXT_LENGTH + 1);
```
Comment (magnumripper) [inline src/opencl_bitlocker_fmt_plug.c]: This doesn't look right. PLAINTEXT_LENGTH is the max length of a password *candidate*. The figure needed here is the maximum length ever needed to represent a "hash" including the ending NULL byte.

File: src/opencl_bitlocker_fmt_plug.c [hunk 2]
```diff
@@ -33,26 +31,22 @@ john_register_one(&fmt_opencl_bitlocker);
...
+//#define FORMAT_NAME             ""
 #define ALGORITHM_NAME          "PBKDF2-SHA256 AES OpenCL"
-#define FORMAT_TAG              "$bitlocker$"
+//#define FORMAT_TAG              "$bitlocker$"
 #define FORMAT_TAG_LEN       (sizeof(FORMAT_TAG)-1)
 #define SALT_ALIGN		1
 #define BITLOCKER_JTR_HASH_SIZE 45
-#define BITLOCKER_JTR_HASH_SIZE_CHAR 77
-#define MIN_KEYS_PER_CRYPT  1   /* These will change in init() */
+#define PLAINTEXT_LENGTH 208
```
Comment (magnumripper) [inline src/opencl_bitlocker_fmt_plug.c]: Our core code currently accept a PLAINTEXT_LENGTH of max. 125 (not including the trailing NULL)

**Comment by e-ago** [comment]
Changed. Next patch I will increase the maximum size of an input password, use more bitlocker_common functions and generally improve the code.

**Comment by magnumripper** [comment]
Good stuff. Please also change ALGORITHM_NAME to "SHA256 AES OpenCL" as discussed above.

**Comment by magnumripper** [comment]
Oh BTW it's some variant of AES, isn't it? The Hashcat issue says AES-XTS and that it may be changed with a registry setting although that will be very rare IRL. If that's correct at all, I presume you only support AES-XTS and ALGORITHM_NAME should be "SHA256 AES-XTS OpenCL".

**Comment by magnumripper** [comment]
@solardiz wrote:
> Also, "SHA-256" is normally spelled with the dash in it (the dash gets dropped in "HMAC-SHA256", because that's how others spell those HMACs, although I think we're now inconsistent about that).

Apparently we're mostly using SHAxxx without the dash in ALGORITHM_NAME. That makes some sense since we're often ending up with a very long line, so I think we should keep doing so and perhaps even change the few ones that have a dash.

**Comment by e-ago** [comment]
@magnumripper ALGORITHM_NAME changed; I'm using a basic AES. The use of modified AES-XTS does not affect the level at which my attack works. I've already tested my implementation with Windows 10 disk images encrypted by myself and with tests offered by kholia and everything works fine.

**Comment by magnumripper** [comment]
Great. I'm merging this now, just open a new PR for future stuff!

**Pull request merged**